### PR TITLE
System Design

### PR DESCRIPTION
**Pull request by tylerjw**
## Description

Resolves #issue-number

Here is my first attempt at a system design.
I broke the system into three modules: Hardware, Game, and Player.

The Hardware system is responsible for producing Game event messages from hardware events (exposure, treatment).
Game event messages are also processed by the hardware to show animations that represent state of the game.
The Game system is responsible for receiving the Game messages and calling functions in the Player module to update the game state and producing Game event messages for the hardware.
The player defines the structure used to hold the state of the game and the functions to transform that state based on different events.

## Testing

This is based on the existing software. I did change the callbacks currently being used to be messages in a different queue because I think it will make implementing simulation easier.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -25,3 +25,56 @@ sudo chmod 666 /dev/ttyUSB0
...
 ```
+# System Design
+
+## Modules
+Communication between the Hardware and Game Event Loop modules is done via message queue.
+Hardware events are placed in the queue by the Hardware module then the Game is ticked, this causes the Game to process the events, using the Player Engine to update the state of the game and enqueues Game messages into a queue for the Hardware system to process (resulting in animations to confirm actions were taken).
+The Game Event Loop uses functions in the player engine to update the state of the game.
+```mermaid
+graph
+    A[Hardware] --> B[(Game Message Queue)]
```
Comment (griswaldbrooks) [inline README.md:36]: Are some of these arrows pointing in the wrong direction?

File: README.md [hunk 2]
```diff
@@ -25,3 +25,56 @@ sudo chmod 666 /dev/ttyUSB0
...
+### Persistent player state
+The persistent game state is stored in special hardware memory represented by a struct from the Player Engine.
+When the Hardware ticks the Game it sends the current player state.
+Then the Game calls functions from the Player Engine to transform the state based on the events it received from hardware.
+Once all messages are processed the persistent state is passed back to the hardware and the new state is stored in the special memory before sleep.
+
+### Hardware Module
+The hardware subsystem performs these steps on each wake-up:
+```mermaid
+graph TD
```
Comment (griswaldbrooks) [inline README.md:52]: Is it relevant here to note https://github.com/kestrel-labs/superspreader/blob/main/arduino/health_monitor/health_monitor.ino#L240 or do you think that's just an implementation detail?

File: README.md [hunk 3]
```diff
@@ -25,3 +25,56 @@ sudo chmod 666 /dev/ttyUSB0
 sudo chown $USER /dev/ttyUSB0
 sudo adduser $USER dialout
 ```
+# System Design
+
+## Modules
+Communication between the Hardware and Game Event Loop modules is done via message queue.
+Hardware events are placed in the queue by the Hardware module then the Game is ticked, this causes the Game to process the events, using the Player Engine to update the state of the game and enqueues Game messages into a queue for the Hardware system to process (resulting in animations to confirm actions were taken).
+The Game Event Loop uses functions in the player engine to update the state of the game.
```
Comment (griswaldbrooks) [inline README.md:33]: Can I suggest
```suggestion
Communication between the Hardware and Game Event Loop modules is done via message queue.
Hardware events are placed in the queue by the Hardware module then the Game is ticked.
The Game then processes the events, using the Player Engine to update the state of the game.
The Player Engine also executes context dependent closures passed to it while processing events, such as playing animations or setting global state.
```

Unless I'm misunderstanding something?

File: README.md [hunk 4]
```diff
@@ -25,3 +25,56 @@ sudo chmod 666 /dev/ttyUSB0
...
+graph
+    A[Hardware] --> B[(Game Message Queue)]
+    B --> C[Game Event Loop]
+    C --> D[(Hardware Message Queue)]
+    D --> A
+    C -..-> E[Player Engine]
+```
+
+### Persistent player state
+The persistent game state is stored in special hardware memory represented by a struct from the Player Engine.
```
Comment (griswaldbrooks) [inline README.md:44]:
```suggestion
To conserve power, the system goes to deep sleep after processing events. To allow game state to persist, it is stored in special hardware memory (called RTC) and represented as a struct from the Player Engine.
```

File: README.md [hunk 5]
```diff
@@ -25,3 +25,56 @@ sudo chmod 666 /dev/ttyUSB0
...
+When the Hardware ticks the Game it sends the current player state.
+Then the Game calls functions from the Player Engine to transform the state based on the events it received from hardware.
+Once all messages are processed the persistent state is passed back to the hardware and the new state is stored in the special memory before sleep.
```
Comment (griswaldbrooks) [inline README.md:47]:
```suggestion
Once all messages are processed the persistent state is passed back to the hardware and the new state is stored in RTC memory before sleep.
```

File: README.md [hunk 6]
```diff
@@ -25,3 +25,56 @@ sudo chmod 666 /dev/ttyUSB0
...
+    E --> F
+    D -->|No| F[Tick Game]
+    F --> Z{Game Msgs?}
+    Z -->|Yes| G[Animate Game Msgs]
+    G --> H(Sleep)
+    Z -->|No| H
+```
+
+### Game Module
+The Game event loop processes the messages placed in the queue from the hardware on each tick.
```
Comment (griswaldbrooks) [inline README.md:66]: So I guess this game messages thing is where we want to go?
Presumably the `game_update` would return/update this queue?
Which then mean there would be a hw engine to process them

**Review by griswaldbrooks — CHANGES_REQUESTED** [review]

File: README.md (hunk 1 above)
Comment (tylerjw) [reply README.md:37]: They might be. The point of this was just to show the different modules and what their interfaces to each other are.

**Review by griswaldbrooks — APPROVED** [review]

**Pull request merged**